### PR TITLE
[Service Bus] Live Tests: Stabilization and Resilience (Part 3)

### PR DESCRIPTION
**Pull request by jsquire**
## Summary

Some of the Service Bus client library Live tests have a measure of non-determinism in them due to timing issues, differences in test runner behavior between the full .NET framework and the .NET Core runner, and what are believed to be some race conditions in asynchronous tests around shared diagnostics and activity tracing.  

This iteration of test stabilization and resilience work was focused primarily on revising some of the asynchronous and threading aspects, isolating diagnostics information to the scope of a particular test, and increasing the isolation between tests themselves, attempting to ensure that each test is entirely self-contained and not subject to environmental considerations or state shared with other tests.

## Outcome

After this set of changes, it is believed that:
- The tests should pass nightly runs consistently,  
- Tests may now be run fully parallel, both within the test suite and across multiple target platforms at once.

## Goals
- Increase reliability of the tests during nightly runs, eliminating timing-related failures and adding stability to tests whose results are non-deterministic.

- Isolate test context such that each test relies only on itself and resources which it directly controls; tests should not be influenced by shared environmental factors or shared state.

- Continue to adding polish to test infrastructure, ensuring that tests have a common base of utilities to work with, and which can clearly and easily be identified as such. 

- Attempt to minimize changes made, keeping the original approach, test cases, code styles

## Non-Goals
- Refactoring of the general approach for testing, the test cases, code style, and other artifacts of the codebase; changes should be kept as tightly scoped and as minimal as possible.

- Adding of additional test cases or new testing scenarios.

### General
  - Created infrastructure for managing common task operations, allowing a consistent approach
    and clean interface for applying timeouts to tasks.

  - Removed usage of `ManualResetEvents` within asynchronous tests to remove potential for thread pool issues, 
    changing over to using `TaskCompletionSource` with asynchronous continuations.

  - Added configuration for the test runner, enabling diagnostic information and allowing tests
    runs in parallel.

  - Added thread safety to the random number generation used within the `ServiceBusScope` class.

  - Minor cleanup to various areas to improve code quality, consistency, and readability.

### Management Client Tests
  - Removed the pattern of creating a class-level management client within each test and relying
    on the `IDisposible` to clean up those instances; tests are now explicitly creating individual clients 
    and responsible for managing their lifecycle.

  - Removed blocking calls on client clean-up, utilizing them in the intended asynchronous manner.

### Diagnostic Tests and Infrastructure
  - Removed the pattern of creating class-level messaging entities within each test and relying
    on the` IDisposible` to clean up those instances; tests are now explicitly creating individual  
    entities and responsible for managing their lifecycle.

  - Removed blocking calls on client clean-up, utilizing them in the intended asynchronous manner.

  - Grouped the diagnostic tests into a single collection, defining a unit of test execution where each is 
    run in isolation, serially.  This prevents the diagnostics information triggered for other tests from interfering 
    with the diagnostics tests, while allowing all but this collection to continue to run in parallel.

  - Relocated diagnostics tests meant to validate features of the `FakeDiagnosticsListener` class into their own
    dedicated test class, rather than intermingling them with tests meant to validate Service Bus client
    library diagnostic features.

## Last Upstream Rebase

Friday, April 05, 2019 2:47pm (EDT)

## Related and Follow-Up Issues

- [[Service Bus Client] Stabilize Live Tests](https://github.com/Azure/azure-sdk-for-net/issues/5446) (#5446)
- [[Service Bus Client] Refactor Diagnostic Tests](https://github.com/Azure/azure-sdk-for-net/issues/5692) (#5692)

### CODE REVIEW

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/xunit.runner.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "$schema": "https://xunit.github.io/schema/current/xunit.runner.schema.json",
+    "longRunningTestSeconds": 61,
+    "diagnosticMessages": true,
+    "parallelizeAssembly": true
```
Comment (jsquire) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/xunit.runner.json]: @mikeharder:  This enables test parallelization within the XUnit runner.   The test suite went from ~21 minutes to ~3 minutes on average and seems to be holding up well in repeated local testing with the .NET core runner.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs [hunk 2]
```diff
@@ -6,558 +6,558 @@ namespace Microsoft.Azure.ServiceBus.UnitTests.Diagnostics
...
 
     public sealed class QueueClientDiagnosticsTests : DiagnosticsTests
     {
-        private QueueClient queueClient;        
-        private bool disposed = false;
-
-        [Fact]
-        [LiveTest]
-        [DisplayTestMethodName]
-        async Task EventsAreNotFiredWhenDiagnosticsIsDisabled()
```
Comment (jsquire) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs]: Note:  This test is an example of one that is not testing diagnostics in the Service Bus client, but rather validating the filtering approach by the test utility `FakeDiagnosticsListener`.   I've relocated these to a dedicated class to try and better differentiate the context being validated.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/MessageTests.cs [hunk 3]
```diff
@@ -140,9 +140,9 @@ public async Task MessageWithMaxMessageSizeShouldWorkAsExpected()
                     var queueClient = new QueueClient(TestUtility.NamespaceConnectionString, queueName, ReceiveMode.PeekLock);
 
                     try
-                    {
-                        var maxMessageSize = (256 * 1024) - 58;     // 58 bytes is the default serialization hit.
-                        var maxPayload = Encoding.ASCII.GetBytes(new string('a', maxMessageSize));
+                    { 
+                        var maxMessageSize = (256 * 1024) - 77;     // 77 bytes is the current serialization hit.
```
Comment (jsquire) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/MessageTests.cs]: Unsure why, but I was seeing fairly consistent failures here after some of the asynchronous changes.   I validated the sizes a few different times and a few different ways, and I was consistently seeing the serialization size as `77` and failing with the original code.   The reported max size of the AMQP link was reported consistently in the failure messaging, so I don't believe that there was a buffering or saturation issue.    If anyone has a thought on why this may be, I'd be grateful for the insight.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        /// <summary>
+        ///   Executes a task without waiting for or observing its completion.
+        /// </summary>
+        /// 
+        /// <typeparam name="T">The type of return value of the task.</typeparam>
+        /// 
+        /// <param name="instance">The target of this fire-and-forget.</param>
+        /// <param name="errorAction">An action to execute should an exception occur.</param>
+        /// 
+        public static async void FireAndForget<T>(this Task<T> instance, Action<Exception> errorAction = null)
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs]: I strongly suggest that we should avoid `async void` methods, you might know that they are not allowed to throw but someone might copy them and modify in a way that would allow throwing and cause very hard to diagnose test crashes. `_ = obj.DoSomethingAsync();` accomplishes the same thing.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs [hunk 5]
```diff
@@ -53,7 +63,7 @@ protected void AssertSendStart(string entityName, string eventName, object paylo
         }
 
         protected void AssertSendStop(string entityName, string eventName, object payload, Activity activity, Activity sendActivity, int messageCount = 1)
-        {
+        {
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: nit: space

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs [hunk 6]
```diff
@@ -130,7 +130,8 @@ async void RunOperationShouldReturnImmediatelyIfRetryIntervalIsGreaterThanOperat
             var policy = RetryPolicy.Default;
             var watch = Stopwatch.StartNew();
             await Assert.ThrowsAsync<ServiceBusException>(async () => await policy.RunOperation(
-                    () => throw new ServiceBusException(true, string.Empty), TimeSpan.FromSeconds(8)));
+                    () => throw new ServiceBusException(true, string.Empty), TimeSpan.FromSeconds(8)))
+                .ConfigureAwait(false);
```
Comment (jsquire) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs]: I had removed this in a previous PR.   Despite it seeming to be an innocuous change, restoring here as there are quite a few areas of the tests that rely on not restoring the context.   I'm open to suggestion, should anyone feel that it's better left gone.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/TransactionTests.cs [hunk 7]
```diff
@@ -83,7 +83,6 @@ public async Task TransactionalSendRollbackTest(bool partitioned, bool sessionEn
                     using (var ts = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled))
                     {
                         await sender.SendAsync(message).ConfigureAwait(false);
-                        ts.Dispose();
```
Comment (jsquire) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/TransactionTests.cs]: If I'm missing something here, please let me know... but it seemed quite useless to call `dispose` the moment before the block scope ends and `dispose` gets called by the `using`.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs [hunk 8]
```diff
@@ -12,30 +12,40 @@ namespace Microsoft.Azure.ServiceBus.UnitTests.Diagnostics
...
 
-        private bool disposed = false;
+        protected ConcurrentQueue<(string eventName, object payload, Activity activity)> CreateEventQueue() =>
+            new ConcurrentQueue<(string eventName, object payload, Activity activity)>();
 
-        internal DiagnosticsTests()
-        {
-            this.events = new ConcurrentQueue<(string eventName, object payload, Activity activity)>();
-            this.listener = new FakeDiagnosticListener(kvp =>
+        protected FakeDiagnosticListener CreateEventListener(string entityName, ConcurrentQueue<(string eventName, object payload, Activity activity)> eventQueue) =>
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: What about disabling parallelization for tests that are using EventListener? Shouldn't make runs much slower as we don't have too many of them but might make things more stable.

Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: To accomplish that you can put tests in common collection with `CollectionDefinition(DisableParallelization = true)]`

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SubscriptionClientDiagnosticsTests.cs [hunk 9]
```diff
@@ -162,5 +125,13 @@ protected void AssertRemoveRuleStop(string entityName, string eventName, object
                 Assert.Equal(removeRuleActivity, activity);
             }
         }
+
+        public void Dispose()
+        {
+            while (Activity.Current != null)
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SubscriptionClientDiagnosticsTests.cs]: Why did you have to add this? Who's leaking activities?

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs [hunk 10]
```diff
@@ -6,558 +6,558 @@ namespace Microsoft.Azure.ServiceBus.UnitTests.Diagnostics
...
-                // let's spin until Complete call starts and ends
-                int wait = 0;
-                while (wait++ < maxWaitSec && this.events.Count < 3)
-                {
-                    await Task.Delay(TimeSpan.FromSeconds(1));
-                }
+                        // message is processed, but abandon happens after that
+                        // let's spin until Complete call starts and ends
+                        int wait = 0;
+                        while (wait++ < MaxWaitSec && eventQueue.Count < 3)
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs]: What about extracting this into a helper?

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SessionDiagnosticsTests.cs [hunk 11]
```diff
@@ -6,213 +6,188 @@ namespace Microsoft.Azure.ServiceBus.UnitTests.Diagnostics
...
+                try
                 {
-                    MessageId = "messageId",
-                    SessionId = sessionId
-                });
-                this.messageSession = await this.sessionClient.AcceptMessageSessionAsync(sessionId);
+                    using (var listener = this.CreateEventListener(queueName, eventQueue))
+                    using (var subscription = this.SubscribeToEvents(listener))
+                    {
+                        listener.Enable();
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SessionDiagnosticsTests.cs]: Seems excessive to have `CreateEventListener` and `SubscribeToEvents` and `listener.Enable` why do we need three of them instead of just one?

Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SessionDiagnosticsTests.cs]: Also CreateEventQueue. Can we return FakeDiagnosticListener instance that would automatically be subscribed to events and have an event queue as a property?

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs [hunk 12]
```diff
@@ -6,558 +6,558 @@ namespace Microsoft.Azure.ServiceBus.UnitTests.Diagnostics
...
 
-        [Fact]
-        [LiveTest]
-        [DisplayTestMethodName]
-        async Task SendAndHandlerFireEvents()
-        {
-            await ServiceBusScope.UsingQueueAsync(partitioned: false, sessionEnabled: false, async queueName =>
-            {
-                this.queueClient = new QueueClient(TestUtility.NamespaceConnectionString, queueName, ReceiveMode.PeekLock);
+                        var processActivity = await tcs.Task.WithTimeout(TimeSpan.FromSeconds(MaxWaitSec));
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs]: Having a `.WithDefaultTimeout()` overload might be useful, usually, people keep copying the same constant everywhere anyway.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs (hunk 6 above)
Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs]: > few areas of the tests that rely on not restoring the context

What relies on that?

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/TransactionTests.cs (hunk 7 above)
Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/TransactionTests.cs]: It might have indirectly tested that calling dispose multiple times doesn't throw but I doubt it. Removing seems fine.

**Review by pakrym — CHANGES_REQUESTED** [review]

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/SubscriptionClientTests.cs [hunk 13]
```diff
@@ -295,22 +295,12 @@ public async Task GetRulesTestCase()
...
-                    {
-                        await subscriptionClient.RemoveRuleAsync(sqlRuleName);
-                        await subscriptionClient.RemoveRuleAsync(correlationRuleName);
-                    }
-                    catch (Exception)
-                    {
-                        // Ignore the exception as we are just trying to clean up the rules that we MIGHT have added.
-                    }
+                    // Attempt to cleanup rules that may or may not exist; ignore any exceptions, as they're expected.
+                    try { await subscriptionClient.RemoveRuleAsync(sqlRuleName); } catch {}
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/SubscriptionClientTests.cs]: _ = Task.WhenAll(...);

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs (hunk 6 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs]: Since a single test often serves as a producer and consumer of messages, they've got the message sender working in the background and typically the receiver working on the original context.   I made the mistake of removing some before I understood what the intent was and things blew up rather spectacularly.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs (hunk 4 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs]: In a normal context, I'd agree.   This extension is explicitly intended to allow the task to complete unobserved while providing a means to discover and act on any exceptions that arise in its execution.   

If I understand your suggestion,  The missing component there is that the task for `DoSomethingAsync` will continue unobserved but lacks the ability to report on exceptions.  Given that it's not being returned, I'm not sure I follow why that would be a better pattern.   Perhaps I'm missing something, could you expand on what you're thinking here, @pakrym?

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs (hunk 8 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: The problem is that ANY activity in the Service Bus library causes diagnostic events and activities to flow.   You'd essentially have to put every single Live test in that collection, removing parallelization.   That was the original position for the tests and there were still some intermittent failures because of diagnostic activities crossing test boundaries.  Adding isolation was a focus to remove those inconsistencies, and had the nice side-effect of allowing parallel tests - which took the overall time from ~21 minutes to ~3 minutes for a full live test run.

The only way to isolate these that I found is to scope a specific listener to only the entities used by a specific test.   That said, I've limited hands-on with the framework diagnostics source/listeners, so you may have a trick or three up your sleeve that I'm not aware of.   Happy to discuss further if you've got thoughts.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SubscriptionClientDiagnosticsTests.cs (hunk 9 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SubscriptionClientDiagnosticsTests.cs]: That should not be there.   It was an intermittent step in removing the base class `Dispose` logic which I was able to eliminate by isolating the listener scope.  Thanks for the catch, this is going away!

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs (hunk 10 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/QueueClientDiagnosticsTests.cs]: Good thought, and I'll take note for the next pass.  This one turned out to be a more substantial set of changes than I anticipated and I'd like to minimize the moving parts.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SessionDiagnosticsTests.cs (hunk 11 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SessionDiagnosticsTests.cs]: This largely kept the same pattern from the existing code.  Collapsing the event queue into the listener is a good idea.  I'm on the fence with automatically subscribing.  The `Enable` method is poorly named (but prevalent in the existing code semantics) - it actually changes the filtering predicate for whether the listener is passed an activity or not.    By default, it filters everything out.

I'm inclined to leave this as the tests have very specific assumptions around what activities are present and their order.   While it is theoretically safe to refactor this and apply a different pattern, I'm not seeing it as a high value gain for the potential risk and effort.

Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/SessionDiagnosticsTests.cs]: After looking over the structure, I'm inclined to leave this as-is.   The responsibility of the listener is determining if an activity is interesting to its consumers and, if so, invoking the delegate that takes action on the observation.   It does not have an understanding of what the consumer of that listener is doing.   

Were I writing this from scratch, I'd likely have gone with a more specialized listener that had more opinionated behavior and could collapse things down close to what you're suggesting.  In this case, however, I don't feel that there is enough benefit to refactoring to warrant the time and risk.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs (hunk 6 above)
Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs]: Is the receiver blocking? If it's async there should be no problem for subsequent continuations to get back on the original context.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+                    return;
+                }
+            }
+
+            // A timeout occured.  Perform the needed actions to request
+            // cancellation, allow the task to complete unobserved, and to
+            // signal the caller.
+
+            cancellationToken?.Cancel();
+            instance.FireAndForget();
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs]: Why do we need FireAndForget here at all? This task is already running or finished.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs (hunk 4 above)
Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs]: `async void` is not something worth using even in a strong team, there are no benefits to it, it just produces patterns that have to be fixed later. FireAndForget with error callback is one of them.

> If I understand your suggestion correctly, the missing component that I see is that, in the proposed approach, the task for DoSomethingAsync will continue unobserved but lacks the ability to report on exceptions. Given that it's not being returned, I'm not sure I follow why that would be a better pattern. Perhaps I'm missing something, could you expand on what you're thinking here, @pakrym?

If we want to observe the exception we shouldn't fire and forget but rather just await the task as part of the test. If we want don't care about the exception assigning task to _ would do it. Do we use error reporting callback anywhere in this change?

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs (hunk 8 above)
Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: I was under the impression that there only parts of live tests use EventListener. If all of them do then you are right and there is no point in limiting the parallelism.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs (hunk 4 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs]: We'll have to agree to disagree on that point.   That said, looking over the usage, it's only in one spot outside this class, so I'll concede that it's easy enough to work around in this case, albeit with less clear semantics, and we can save the larger philosophical discussion for another time.    :)

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs (hunk 8 above)
Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: I looked again, there is a ton of live tests that are not using the diagnostic listener. Running all of them in parallel if completely fine, `[CollectionDefinition(DisableParallelization = true)]` would force diagnostic tests to run after all other tests were run in parallel.

Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: Are we talking about two different things here?   I'm noting that the Service Bus library being tested is producing diagnostic events that flow through the listener.   Even though only the diagnostics family of tests make use of the listener, without the filtering that I put in to keep a listener focused on activities for a specific entity, the listener would be grabbing data from all activities happening in the other tests.

I feel like I'm missing your point here.  What are you proposing, exactly?   I thought it was removing the entity filtering, but ....

Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: > the listener would be grabbing data from all activities happening in the other tests.

Why if there would be no other tests running in parallel? 

>  What are you proposing, exactly? I thought it was removing the entity filtering, but ...

Moving diagnostics test into a collection so they won't run in parallel with anything else or each other and removing the filtering 

Maybe I'm missing something :)

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs (hunk 6 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs]: They're interleaving delays, multiple senders, multiple receivers for different scenarios.   Intuitively, I expected it to be safe to remove, but the reality was very different.  I honestly didn't see justification to spend the time to really dig into the details of why they went with the pattern that they did.  You're welcome to dig deeper and file issues to refactor things, but I'm going to call it out of scope for now.

Comment (pakrym) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/RetryTests.cs]: Sounds good.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs [hunk 15]
```diff
@@ -151,7 +99,7 @@ public static async Task<T> WithTimeout<T>(this Task<T> instance, TimeSpan timeo
             // signal the caller.
 
             cancellationToken?.Cancel();
-            instance.FireAndForget();            
+            var _ = instance.ConfigureAwait(false);
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs]: Why is this required?

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs [hunk 16]
```diff
@@ -100,8 +48,8 @@ public static async Task WithTimeout(this Task instance, TimeSpan timeout, Cance
             // cancellation, allow the task to complete unobserved, and to
             // signal the caller.
 
-            cancellationToken?.Cancel();
-            instance.FireAndForget();
+            cancellationToken?.Cancel();
```
Comment (pakrym) [inline src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs]: nit: space

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs (hunk 8 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: I think that you may be.   The listener is an instance, not shared among the tests.   Each instance is filtering diagnostic information to only the entities that it's owner cares about.   However, the source of diagnostics is global.  Any operation that uses the Service Bus client library potentially produces diagnostic information.   That information is pushed to every listener instance that is an observer, with a predicate called to allow filtering.    The underlying problem wasn't that listeners were being shared by the tests, it was that each listener is observing the global diagnostics and was seeing information produced by other callers in all other tests - whether they have a listener or not.  The producer is still creating diagnostics for them and sending it to everyone.

For a fun illustration, lets say that every 3 seconds, a different person smacks me in the head.  You and Alex are watching me get smacked.  You only care about people with blond hair that smack me but you see every single smack because there's only one me, and I'm being beaten in public.    Your only choice is to teach yourself to ignore it when someone who isn't blond hits me.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs (hunk 15 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Infrastructure/TaskExtensions.cs]: yeah... fair point.  That's already running.  My bad.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs (hunk 8 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: Pavel and I discussed offline;  I wasn't fully understanding his suggestion.   Going to give it a look and see what the net effect is on overall stability.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/xunit.runner.json (hunk 1 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/xunit.runner.json]: Note: with the change to run diagnostics tests in isolation and serially, the suite run time is ~4 minutes now.

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs (hunk 8 above)
Comment (jsquire) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/Diagnostics/DiagnosticsTests.cs]: After some investigation, this seemed to work out well.   I've converted the approach to remove filtering by entity and, instead, group diagnostic tests into a collection to be run as a single unit of execution.  They'll run in isolation with each diagnostic test done serially to ensure that there are no other consumers of the client generating diagnostics information that may interfere.

Note that I did keep the logic in the `DiagnosticTests` base class that identifies mismatches between the entity a test is interested in and activities for a different target entity.  Instead of filtering, it now emits a log entry.  This information was incredibly useful in diagnosing the test instability, so I believe it worth the additional complexity and time to execute.

**Review by pakrym — APPROVED** [review]

File: src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/xunit.runner.json (hunk 1 above)
Comment (mikeharder) [reply src/SDKs/ServiceBus/data-plane/tests/Microsoft.Azure.ServiceBus.Tests/xunit.runner.json]: This is great, I think <= 5 minutes is a good goal for a live test suite.

**Review by mikeharder — APPROVED** [review]
